src/program-editors.rst

```diff
@@ -22,6 +22,127 @@ Visual Programming Languages
+
+Applications
+~~~~~~~~~~~~
+
+LabVIEW
+^^^^^^^
+
+LabVIEW (Laboratory Virtual Instrument Engineering Workbench) is a system-design platform and development environment for a visual programming language designed by National Instruments based on visual data-flow language paradigm.
+LabVIEW followed the convention of a box-line structure and enriched it with constructs for conditionals, loops and sequences. And different types of lines represent different types of data (numbers, booleans, strings, arrays ...) 
+flowing among boxes. The operators provided perfectly covered almost all needs to construct a virtual electronic instrument operating in real time but make considerably less coverage provision for other domains of computing.
+Besides the usability, LabVIEW has been keeping a great contribution on researches about cognitive effects of visual programming representations.
+
+.. container:: bib-item
+
+  .. bibliography:: notation.bib
+    :filter: key == 'vp2001labview'
+
+This article reports the results of an exploratory survey of experienced LabVIEW
+users, a main goal being to learn how LabVIEW programmers think that the visual
+representation of LabVIEW affects their programming.
+
+Prograph
+^^^^^^^^
+
+Prograph is a visual data flow and object-orientated language marketed by TGS Systems.
+The architecture comprises methods (subroutines) represented
+as icons and connected by the usual lines, carrying data objects from top to bottom of
+the screen. A method browser is used to browse methods and cases displayed in a seperate window.
+The dataflow lines cannot contain bends but can be placed diagonally which is different from LabVIEW.
+
+Luna
+^^^^
+
+Luna is a textual and visual functional programming language, which provides a data processing and visualization environment which is able to immediately 
+connect programmers to what they are building. Instead of being hard coded, components in Luna are simply nested and editable data flow graphs which allows immersive programming experience.
+
+
+others
+^^^^^^
+
+.. container:: bib-item
+
+  .. bibliography:: notation.bib
+    :filter: key == 'gui2017vpl'
+
+This article introduced a very interesting invention based on visual dataflow programming paradigm. The invention is about method for adapting GUI-based instrument components in a visual dataflow programming language.
+
+Drawbacks
+~~~~~~~~~
+
+Scale issue
+^^^^^^^^^^^
+
+Given the fact that visual representations of programming languages can usually manifest as much detail of a program as possible, a new concern will automatically arise when one is trying to program in a large scale using 
+a visual data flow programming language, when dataflows become very complex and hierarchical.
+
+.. container:: bib-item
+
+  .. bibliography:: notation.bib
+    :filter: key == 'vp2017lvss'
+
+This article describes a visual programming environment
+that we are extending toward a visual software engineering
```

Comment:
"we are extending" who is we? if you copied this text from the paper, you need to put it in quotation marks, otherwise it is plagiarism.